docs/architecture/adr-062-p2p-architecture.md

```diff
@@ -0,0 +1,531 @@
+type Endpoint struct {
+    // Protocol specifies the transport protocol, used by the router to pick a
+    // transport for an endpoint.
+    Protocol Protocol
+
+    // Path is an optional, arbitrary transport-specific path or identifier.
+    Path string
+
+    // IP is an IP address (v4 or v6) to connect to. If set, this defines the
+    // endpoint as a networked endpoint.
+    IP net.IP
+
+    // Port is a network port (either TCP or UDP). If not set, a default port
+    // may be used depending on the protocol.
+    Port uint16
+}
+
+// Protocol identifies a transport protocol.
+type Protocol string
+```
+
+Endpoints are arbitrary transport-specific addresses, but if they are networked they must use IP addresses and thus rely on IP as a fundamental packet routing protocol. This enables policies for address discovery, advertisement, and exchange - for example, a private `192.168.0.0/24` IP address should only be advertised to peers on that IP network, while the public address `8.8.8.8` may be advertised to all peers. Similarly, any port numbers if given must represent TCP and/or UDP port numbers, in order to use [UPnP](https://en.wikipedia.org/wiki/Universal_Plug_and_Play) to autoconfigure e.g. NAT gateways. Note that a concrete design for detection and advertisement of externally-reachable addresses (e.g. via NAT gateways) is out of scope for this ADR, and may be covered in a separate ADR.
+
+Non-networked endpoints (without an IP address) are considered local, and will only be advertised to other peers connecting via the same protocol. For example, an in-memory transport might use `Endpoint{Protocol: "memory", Path: "foo"}` as an address for the node "foo", and this should only be advertised to other nodes using `Protocol: "memory"`.
+
+#### Connections and Streams
+
+A connection represents an established transport connection between two endpoints (and thus two nodes), which can be used to exchange bytes via logically distinct IO streams. Connections are set up either via `Transport.Dial()` (outbound) or `Transport.Accept()` (inbound). The caller is responsible for verifying the remote peer's public key as returned by the connection.
+
+Data is exchanged over IO streams created with `Connection.Stream()` using an arbitrary `StreamID`. These implement the standard Go `io.Reader` and `io.Writer` interfaces to read and write bytes. Transports are free to choose how to implement such streams, e.g. by taking advantage of native stream support in the underlying protocol or through multiplexing.
+
+`Connection` and the related `Stream` interfaces are:
+
+```go
+// Connection represents an established connection between two endpoints.
+type Connection interface {
+    // Stream opens a logically distinct IO stream within the connection,
+    // using an arbitrary stream ID. The stream must be created before it
+    // can receive data, and must be closed before the ID can be reused.
+    Stream(StreamID) Stream
```

Comment:
Having the caller specify the stream ID might be problematic. For example, QUIC will just assign numerically increasing stream IDs, and using a specific ID will cause all lower IDs to also automatically be allocated. This might e.g. require the QUIC transport to map QUIC stream IDs onto transport stream IDs and use an initial handshake to make both nodes agree on the transport stream ID.

Another option here is for the streams to not have IDs at all, and instead have the channel do a handshake when it allocates a stream. This seems cleaner, and would allow the handshake to include additional channel metadata if and when necessary.

Reply:
> instead have the channel do a handshake when it allocates a stream

static IDs makes debugging easier (e.g. you've received msg{StreamID:1} and you know StreamID: 1 is consensus

Reply:
Can we have this be implementation specific where in the case of QUIC it ignores the provided streamID or does some internal mapping?

Reply:
> static IDs makes debugging easier

True. Most code would interact with channels though, which have a fixed ID, only P2P internals would use streams.

> Can we have this be implementation specific where in the case of QUIC it ignores the provided streamID or does some internal mapping?

Yes, it would have to do something like that. But in that case the stream needs to do a handshake so both sides (both the local and remote node) agree on which ID the stream should have, which is a bit unfortunate. Totally workable though.

Reply:
> Most code would interact with channels though, which have a fixed ID, only P2P internals would use streams.

That makes sense to me.

Reply:
Been giving this some more thought, and I think we should drop stream IDs and use channel handshakes. This can be made backwards-compatible with the current MConn protocol by having the MConn transport fake the local channel handshake using the data from the previous connection handshake. Will update the spec after the dev call.